rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp

```diff
@@ -81,7 +81,70 @@ class MultiThreadedExecutor : public rclcpp::Executor
 private:
   RCLCPP_DISABLE_COPY(MultiThreadedExecutor)
 
-  std::mutex wait_mutex_;
+  /// \internal A mutex that has two locking mechanism, one with higher priority than the other.
+  /**
+   * After the current mutex owner release the lock, a thread that used the high
+   * priority mechanism will have priority over threads that used the low priority mechanism.
+   */
+  class MutexTwoPriorities
```

Comment:
Honestly, this doesn't need any special access to the `MultiThreadedExecutor` class, so it should be in its own file and and it should also use a `.cpp` file.

Reply:
You could put it in a detail namespace too, like was done here:

https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/wait_set_policies/detail/write_preferring_read_write_lock.hpp

Reply:
I defined `MultiThreadedExecutor` as a nested class only because it was a private implementation detail I don't think we're going to use anywhere else.
Moving it to its own header/cpp in the detail namespace sounds good to me.

Reply:
done in https://github.com/ros2/rclcpp/pull/1516/commits/7b9548d583f9f822809dc0d8b12fa60997b1e2e9 and https://github.com/ros2/rclcpp/pull/1516/commits/6618464239ab6492a9f172cb670f7075c4f4c0b1